### DISCUSSION TITLE
Environment Cleanup

### DISCUSSION THREAD

**Discussion by waldo1001** [microsoft/alguidelines#163] [BC Best Practices]
A guideline to implement "Environment Cleanup" of your module.  There is little to no documentation about it on Docs, but needs proper attention

**Comment by curateideas** [comment]
What do you have in mind when the "cleanup" should happen?

**Reply by waldo1001** [reply]
When an environment is copied to a new one

**Reply by curateideas** [reply]
@waldo1001 : While reading you blog post, I had a small deja vu. I stumbled upon the cleanup of settings in the Microsoft Docs a few weeks ago when preparing a presentation about the restore-function in BC Admin Center. And there they mentioned the exact same setups:

![image](https://user-images.githubusercontent.com/58227881/159888511-88bd96b3-380e-45cd-beba-3d4e8d3163ba.png)

https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/administration/tenant-admin-center-backup-restore#considerations-and-limitations

So it seems Microsoft uses them in case of **restoring** an environment (which seems to be a sort of 'copy' of an environment, too).

**Comment by PeterConijn** [comment]
I think this is an important, yet often overlooked, process in these copy procedures. I am linking your (@waldo1001)  post about this on Dynamics Community, since it has important information: [link](https://community.dynamics.com/blogs/post/?postid=346240b6-5e90-4e46-b041-a6dd9f454bcf)

Like Waldo said in that post, the environment triggers (most likely) only work in SaaS. For OnPrem, we intend to integrate this in the SQL Copy script we use for copying, but since we live in a multi-app world, this will need some checks on whether an app is installed.

Either method will also need maintenance if new tables (that fall under the "must clean/disarm" scenario) are added. This means that we will need to raise awareness in our feature write-ups (like we have for "does this need install/upgrade code"?) and development teams too.
